File: utbot-framework/src/main/java/org/utbot/quickcheck/internal/GeometricDistribution.java


package org.utbot.quickcheck.internal;

import org.utbot.quickcheck.random.SourceOfRandomness;

import static java.lang.Math.ceil;
import static java.lang.Math.log;

public class GeometricDistribution {
    public int sampleWithMean(double mean, SourceOfRandomness random) {
        return sample(probabilityOfMean(mean), random);
    }

    int sample(double p, SourceOfRandomness random) {
        ensureProbability(p);

        if (p == 1)
            return 0;

        double uniform = random.nextDouble();
        return (int) ceil(log(1 - uniform) / log(1 - p));
    }

    double probabilityOfMean(double mean) {
        if (mean <= 0) {
            throw new IllegalArgumentException(
                "Need a positive mean, got " + mean);
        }

        return 1 / mean;
    }

    private void ensureProbability(double p) {
        if (p <= 0 || p > 1) {
            throw new IllegalArgumentException(
                "Need a probability in (0, 1], got " + p);
        }
    }
}

File: utbot-greyboxfuzzer/src/main/kotlin/org/utbot/greyboxfuzzer/quickcheck/internal/GeometricDistribution.kt
package org.utbot.greyboxfuzzer.quickcheck.internal

import org.utbot.greyboxfuzzer.quickcheck.random.SourceOfRandomness
import kotlin.math.ceil
import kotlin.math.ln

class GeometricDistribution {
    fun sampleWithMean(mean: Double, random: SourceOfRandomness): Int {
        return sample(probabilityOfMean(mean), random)
    }

    fun sample(p: Double, random: SourceOfRandomness): Int {
        ensureProbability(p)
        if (p == 1.0) return 0
        val uniform = random.nextDouble()
        return ceil(ln(1 - uniform) / ln(1 - p)).toInt()
    }

    fun probabilityOfMean(mean: Double): Double {
        require(mean > 0) { "Need a positive mean, got $mean" }
        return 1 / mean
    }

    private fun ensureProbability(p: Double) {
        require(!(p <= 0 || p > 1)) { "Need a probability in (0, 1], got $p" }
    }
}